### DISCUSSION TITLE
systemd graceful shutdown

### DISCUSSION THREAD

**Discussion by ririsoft** [keycloak/keycloak#10180] [Archive]
Hello,

I am using a basic systemd unit service file for keycloak 17.0.0 quarkus distribution on Debian 11 :

```sh
# /etc/systemd/system/keycloak.service
[Unit]
Description=The WildFly Application Server
After=syslog.target network.target postgresql.service
Before=httpd.service
PartOf=postgresql@13-main.service

[Service]
User=keycloak
UMask=0027
ExecStart=/opt/keycloak/keycloak-17.0.0/bin/kc.sh start 

[Install]
WantedBy=multi-user.target
```

But shutdown is always in systemd failed status with exit code 143.
Logs are not helping:

```sh
févr. 13 15:38:49 horous kc.sh[493274]: 2022-02-13 15:38:49,948 INFO  [org.infinispan.CLUSTER] (Shutdown thread) ISPN000080: Disconnecting JGroups channel `ISPN`
févr. 13 15:38:50 horous kc.sh[493274]: 2022-02-13 15:38:50,002 INFO  [io.quarkus] (Shutdown thread) Keycloak stopped in 0.113s
```

Do you have any best practice for shutting down keycloak.x properly with systemd, please ?
Thank you very much in advance for your help here.

**Comment by tgurr** [comment]
Experiencing the same problem, I'm not sure if this is the correct solution or rather just a dirty workaround but it seems to be quite common in other systemd services files in java based software like e.g. openhab, adding:

`SuccessExitStatus=0 143` to the `keycloak.service` works around the problem.

```
Feb 14 13:20:21 localhost kc.sh[174020]: 2022-02-14 13:20:21,725 INFO  [io.quarkus] (Shutdown thread) Keycloak stopped in 0.049s
Feb 14 13:20:21 localhost systemd[1]: keycloak.service: Deactivated successfully.
Feb 14 13:20:21 localhost systemd[1]: Stopped keycloak.service - Keycloak server.
Feb 14 13:20:21 localhost systemd[1]: keycloak.service: Consumed 11.066s CPU time.
```

**Reply by ririsoft** [reply]
Indeed this is a workaround, but I was wondering if there is actually an issue on Quarkus/Keycloak side so that 143 is returned instead of 0.

**Reply by pedroigor** [reply]
We should not return `143` but `0` or `1`. The server seems to be shut down by looking at the logs and the status should be 0.

Are you able to do the same thing using the legacy dist without the change you did?

**Reply by tgurr** [reply]
I didn't try 17.0.0 legacy specifically but adding any `SuccessExitStatus` to the systemd unit file at least wasn't necessary with 16.1.1 as far as I can tell.

**Reply by pedroigor** [reply]
Thanks. Created an issue to investigate the behavior.

**Comment by darius-m** [comment]
It may also be connected to [this issue](https://issues.redhat.com/browse/WFCORE-5551). For WildFly, the start script contains a loop that tries to kill Keycloak, but does not properly check if it is still running.

**Comment by pedroigor** [comment]
Looks like the JVM default is to return an exit code of 128+signal-id. E.g for SIGTERM you'd see an exit code of 143 (128+15).

We can check if changing our script solves the issue (assuming Wildfly is working fine with systemd) but I'm wondering if we can document this behavior and provide a guide that shows how to make Keycloak a systemd service.

Would that be enough? I don't see any issue using the `SystemExitStatus` to set the expected exit code(s).

**Reply by ririsoft** [reply]
As long as it is documented and as the project provides an example systemd unit file, I do not believe this in an issue to use `SuccessExitStatus`.

**Reply by pedroigor** [reply]
@ririsoft Let me prepare something based on your example and I'll ask you to review it if you don't mind.

**Reply by ririsoft** [reply]
@pedroigor I will be pleased to help. I have very little experience with running JVM/Quarkus, but I have some experience working with systemd and can help on the hardening part.

**Reply by pedroigor** [reply]
Thanks, created https://github.com/keycloak/keycloak/issues/10357. Will try to come up with something this week.

**Reply by inglor** [reply]
@pedroigor My approach to harden the keycloak systemd service in Arch Linux can be found:
https://github.com/archlinux/svntogit-community/tree/708ece30acb2d1c5fd401b6f5a8169d58d34b443/trunk

Feedback is welcome

This is related to https://github.com/keycloak/keycloak/discussions/10323

**Reply by soundsessential** [reply]
Folks, I have been writing an installation tutorial on Keycloak 17. I have used Certbot and Let's Encrypt certs which live in a protected folder in /etc. If you start Keycloak as root using sudo bin/kc.sh start, everything is great. But if you create a user, give it ownership of the keycloak folder, this user can not access the certs when a systemd file is used. I'm out of my depth here but wanted to pass on that I think adding + in front of the command gives it temporary privileges to get Keycloak running, allowing it to access the certs. I'm not sure if the process is then running as root and this is considered bad? The very simple systemd is below. If you have any comments/tips/concerns, please do let me know. I hope to get this tutorial out in the next day or so. Many thanks and hope this helps with the fine work you are all doing.

```
# /etc/systemd/system/keycloak.service
[Unit]
Description=Keycloak Server
After=syslog.target network.target mysql.service
Before=httpd.service

[Service]
User=keycloak
Group=keycloak
ExecStart=+/opt/keycloak/keycloak-17.0.0/bin/kc.sh start

[Install]
WantedBy=multi-user.target
```

**Reply by darius-m** [reply]
Starting the service using `+` appears to be more than required. You should at least use `!`, which according to `man systemd.service` does not also grant additional capabilities, but alters the `Users`, `Group` and `SupplementaryGroups` parameters. However, I think for certificates (or other files that only keycloak should access but not write), changing ownership to `root:keycloak` and giving `g=rX` permissions to the files and directories would make more sense.

I do not believe that either `+` or `!` grant "temporary" privileges to the command (i.e., the permissions are time limited in any manner), but instead will grant root permissions for that command (in turn making the `kc.sh` script run as root). If you use a service that does this, it should open the files and then fork to create a child process that inherits the open file descriptors, but has limited privileges afterwards. Without checking the code, I think that the nginx (and maybe apache httpd) server behaves like this, where it has a parent process that runs as root, but forks workers to run as a limited user.

**Reply by soundsessential** [reply]
Thank you for the quick reply and suggestions. This is what the certs look like currently in this folder /etc/letsencrypt/live/keycloak.mydomain.com/

```
lrwxrwxrwx 1 root root  52 Feb 21 12:17 cert.pem
lrwxrwxrwx 1 root root  53 Feb 21 12:17 chain.pem
lrwxrwxrwx 1 root root  57 Feb 21 12:17 fullchain.pem
lrwxrwxrwx 1 root root  55 Feb 21 12:17 privkey.pem
-rw-r--r-- 1 root root 692 Feb 21 12:17 README
```

The `!` option does appear to work but as you suggest may not be ideal. From very brief research, I do believe you are right that Apache binds as root and then switches to the lower privileges to run the remaining processes. If we can not currently do this with Keycloak and you suggest the ownership/permissions adjustment, could you spell out the commands you are suggesting to run to do this. I'll happily test and report back for others to learn from. I am concerned about running the incorrect commands and making keys available to who they shouldn't be! Many thanks again for your time.

**Reply by darius-m** [reply]
The following "should" be good enough, but since we will be changing permissions for the files to allow the entire group to read the files, only the `keycloak` user should be in the group. I prefer this approach, since only the owner (or root) can change file permissions; the group cannot.

Ideally, the system (or VM) that runs keycloak should only run this service for the best isolation.

Do keep in mind that the files in the let's encrypt's directory are symlinks, so the actual permissions may not be handled as expected (the directory's permissions are altered on the directory itself, but file permissions are changed on the link's destination, since symlinks do not have permissions themselves - that's why they show as `rwxrwxrwx`).
```bash
# Change ownership of the files so they are owned by root, but the keycloak group can read them
# Add execute for directories, so they can be traversed). Other users cannot access the files at all
chown -R root:keycloak /etc/letsencrypt/live/keycloak.mydomain.com
chmod -R u=rwX,g=rX,o=- /etc/letsencrypt/live/keycloak.mydomain.com
```

Alternatively, you could give ownership to the `keycloak` user, so only that user can access them. However, since the user owns the files this time (as opposed to `root` in the first example), the user can change the permissions themselves. Root will be able to access the files regardless of permissions.
```bash
# Change ownership of the files so they are owned by the keycloak user
chown -R keycloak:root /etc/letsencrypt/live/keycloak.mydomain.com
chmod -R u=rX,g=rX,o=- /etc/letsencrypt/live/keycloak.mydomain.com
```

**Reply by soundsessential** [reply]
Wow, hugely grateful for the response. Unfortunately that does not appear to have worked. I added sudo in front of those commands (the first two which I think is preferred) and they executed without errors. But in using the systemd without `!` or `+` I get this (which is the same as I have been getting - essentially access denied to the certs):

```
Feb 21 16:23:29 keycloak17-ubuntu kc.sh[945]: 2022-02-21 16:23:29,125 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: /etc/letsencrypt/live/keycloak.mydomain.com/fullchain.pem
Feb 21 16:23:29 keycloak17-ubuntu kc.sh[945]: 2022-02-21 16:23:29,128 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Key material not provided to setup HTTPS. Please configure your keys/certificates or start the server in development mode.
```

The cert permissions now look like this:
```
lrwxrwxrwx 1 root keycloak  52 Feb 21 12:17 cert.pem
lrwxrwxrwx 1 root keycloak  53 Feb 21 12:17 chain.pem
lrwxrwxrwx 1 root keycloak  57 Feb 21 12:17 fullchain.pem
lrwxrwxrwx 1 root keycloak  55 Feb 21 12:17 privkey.pem
-rw-r----- 1 root keycloak 692 Feb 21 12:17 README
```

So they did change but the Keycloak user I guess is still being denied access (or does it not have access to something else perhaps?)

For full disclosure, when setting up that user the commands used were:
```
sudo groupadd -r keycloak
sudo useradd -r -g keycloak -d /opt/keycloak -s /sbin/nologin keycloak
cd /opt
sudo chown -R keycloak: keycloak
sudo chmod o+x /opt/keycloak/keycloak-17.0.0/bin/
```

**Reply by darius-m** [reply]
There may be an issue with the fact that you are changing permissions on symlinks (the permissions column starts with `l`), with the actual certificates and keys being in a different directory. It is a bit strange that the owner of the symlink has changed; probably because of this, Keycloak is most likely not able to access the actual certificates. You have to also change permissions on the target files (and possibly parent directories) to allow keycloak to access them.

However, it comes to mind that the recommended (and honestly easier) approach (at least for versions before 17) would be to use a load balancer / reverse proxy, such as HAProxy or nginx. If you use one of them, you can set the permissions back to their defaults (the certificate / key files are probably owned by `root:root` and with permissions of `u=rwX,g=rX,o=-`). The load balancer should by default start as root, read the certificates and then drop to a non-privileged user, so the certificates can have their permissions as restricted as possible. Even if you only have one server, you could have Keycloak start on port `8080` or `8443` that is blocked for outside connections using a firewall, while the reverse proxy listens on `443` and redirects requests to `localhost:8080` or `localhost:8443`. Keycloak itself could use a self signed certificate, or even no certificate at all if communication between the reverse proxy and Keycloak is done locally.

**Reply by soundsessential** [reply]
Yeah a reverse proxy would probably be easier, but our user base is not familiar with Linux for the most part (although getting better at it) so introducing two Linux factors may prevent adoption. So, while I need to test from scratch, as I've been messing with this thing, if I take your commands and up the hierarchy, I think this works:

```
chown -R root:keycloak /etc/letsencrypt
chmod -R u=rwX,g=rX,o=- /etc/letsencrypt
```

This is based on the fact that only one set of keys would be stored here, so no concerns about the keycloak user having access to keys of something-else.mydomain.com.

Any security concerns with doing this? I think renewals would work ok but I would need to test that.

**Reply by soundsessential** [reply]
Arrgh, scratch that. It appears to work but on clicking on a Realm and then Roles (and some other buttons, but not all) you get a blank menu and the page doesn't go where you want it to. You end up on https://keycloak.mydomain.com:8443/admin/master/console/#/notfound

<img width="1091" alt="Screen Shot 2022-02-21 at 12 48 11 PM" src="https://user-images.githubusercontent.com/4053754/155005566-be0d8940-8ac0-4b98-8a63-98fece6e0ce2.png">


I think, until there's further guidance from folks way more smart than me, I'll go with launching with `!`.

@darius-m Thank you, sincerely, for taking the time. Much appreciated.

**Reply by darius-m** [reply]
I think that the best approach would be using a dedicated proxy that can drop privileges. That way, even if the server is partially compromised, extracting the keys from memory or open file descriptors would be a bit harder compared to accessing the keys in the filesystem. Ideally, Keycloak would also be able to do this when certificates are concerned.

However, considering that Keycloak is most of the time a critical service, if someone manages to compromise the service and gain access on the server as the `keycloak` user, where they can access user information or change the server's configuration, a compromised Let's Encrypt key may not be the largest problem.

On a side note, I recommend that you use a similar approach to other files used by Keycloak, where other users cannot access any of the files. All commands that concern Keycloak should be run as the `keycloak` user, regardless of purpose (both administration or SSO web service). Granting the service `root` privileges is not usually recommended; allowing other users read / execute permissions on the files is not necessary.

I personally prefer using file ACLs to make sure new files / directories are not created (by default) with permissions for others, but this makes inspecting permissions quite a bit harder, since ACLs are not displayed by `ls`. If you want to do this, you can run the following commands (two commands are required, since the default permissions are not set immediately). Directories and files that have ACLs set have the last character of the permissions column (that is normally `.`) changed to `+`.
```bash
setfacl -R -d -m o::- /opt/keycloak
setfacl -R -m o::- /opt/keycloak # this is basically the same as "chmod -R o=- /opt/keycloak"
```

ACLs can also be used to grant only the `keycloak` user / group additional permissions on the certificate files, but I strongly believe that this would make permission checking a lot harder, so I cannot recommend it.

Your last error is probably an issue with some other Keycloak files. If the service starts and you can connect to it, the certificates can be accessed, so there should not be any issues there. Please make sure the `keycloak` user can properly access all files under the `/opt/keycloak` directory (from the commands you mentioned above, it should be correct, but there must be an error since you cannot access all components). Also check if you find anything in the logs - you should be able to use `journalctl -xe -u keycloak` or check the `logs` directory.

**Reply by soundsessential** [reply]
Ha ha, that's all so over my head but I again appreciate the effort. So I checked the logs after clicking around a bit and making it fail. This is in case maybe it's a bug but I imagine it's my keycloak user, although I did give it ownership as above:

`Feb 21 18:30:19 keycloak17-ubuntu kc.sh[940]: 2022-02-21 18:30:19,769 WARN  [org.keycloak.encoding.ResourceEncodingProvider] (executor-thread-0) Failed to encode resource: java.io.FileNotFoundException: /opt/keycloak/keycloak-17.0.0/bin/../data/tmp/kc-gzip-cache/9oi5d/admin/keycloak/partials/notfound.html.gz (Permission denied)`

I should probably build this from scratch again to rule out any conflicts I've introduced.

**Reply by soundsessential** [reply]
Yeah so just to confirm, with a clean fresh install, applying permissions as above to the lets encrypt folders again results in that error. It seems to be consistently happening when on the master realm > realm settings, you can get to the first 5 tabs (general thro' themes) but from localizations onwards it results in the error and the blank screenshot. 

I'll leave this with people with more knowledge than myself! I'll have to go with my elevated privileges workaround for now but if anyone finds a solution to allow the keycloak user to access the lets encrypt certs without these issues (or if it's nothing to do with the certs and has a solution to the error), please do let me know. I'd love to push Keycloak in our community but absolutely want to do it in the best possible way. I'll post a link here to the blog post when it's live in case anyone wants to follow it to see what might be up. Thanks all!

**Reply by soundsessential** [reply]
Our tutorial is available. As the work on the systemd unit file moves forward, if anyone has any ideas on why the error above may be happening on a user with ownership of the keycloak folder and permissions to the Let's Encrypt folder (although this is not applied within the tutorial as yet), please do let us know. We're eager to get this updated to not using elevated privileges. Thanks all.
https://www.soundsessential.com/blog/214-keycloak-17-filemaker-installation-configuration-tutorial-part-3-preparing-keycloak-17
https://www.soundsessential.com/blog/215-keycloak-17-filemaker-installation-configuration-tutorial-part-4-starting-keycloak-17-next-steps

**Comment by pedroigor** [comment]
Sorry for the late reply. Looks like we have quite a few material and information to provide instructions on how to run using systemd.

That said, I would like to know from you what are the next steps to make solve this discussion/issue. 

Looks like we can use this article https://www.soundsessential.com/blog/215-keycloak-17-filemaker-installation-configuration-tutorial-part-4-starting-keycloak-17-next-steps as the basis for a new guide in our website.

The remaining issue though is related to figuring out how to avoid running the service with elevated privileges.

Does that make sense?

**Comment by tgurr** [comment]
If it's of interest this is the systemd file we currently use for keycloak >= 17:

```
[Unit]
Description=Keycloak server
After=network-online.target
Wants=network-online.target systemd-networkd-wait-online.service

[Service]
User=keycloak
Group=keycloak
ExecStart=/opt/keycloak/bin/kc.sh start --auto-build
WorkingDirectory=/opt/keycloak
ReadWritePaths=/opt/keycloak/conf /opt/keycloak/data /opt/keycloak/lib/quarkus
SuccessExitStatus=0 143

# Hardening options
CapabilityBoundingSet=
AmbientCapabilities=
NoNewPrivileges=true
ProtectHome=true
ProtectSystem=strict
ProtectKernelTunables=true
ProtectKernelModules=true
ProtectControlGroups=true
PrivateTmp=true
PrivateDevices=true
LockPersonality=true

[Install]
WantedBy=multi-user.target
```

**Reply by soundsessential** [reply]
Are you using SSL certs and if so, where do you save these to? I seem to have had trouble with having Lets Encrypt certs save to their default location and then have a non root user get access to them. 

I think I may have had issues with some of the data paths too, which your ReadWritePaths entry may resolve. I will test this. Thanks for sharing.

**Reply by tgurr** [reply]
> Are you using SSL certs and if so, where do you save these to?

I'm using nginx as a reverse proxy for that.

/opt/keycloak/conf/keycloak.conf

```
# Listen IP, default: 0.0.0.0
http-host=127.0.0.1

# HTTP listener, default: false
http-enabled=true

# The proxy address forwarding mode if the server is behind a reverse proxy.
proxy=edge

# Hostname for the Keycloak server.
hostname=keycloak.domain.com
```

/etc/nginx/nginx.conf
```
    server {
        listen 80;
        listen [::]:80;
        server_name keycloak.domain.com;
        return 301 https://$server_name$request_uri;

        access_log off;
        error_log off;
    }

    server {
        listen 443 http2 ssl;
        listen [::]:443 http2 ssl;
        server_name keycloak.domain.com;

        # SSL security configuration
        include /etc/nginx/ssl_security.conf;

        access_log off;
        error_log off;

        # restrict access to master realm
        location /realms/master/ {
            proxy_pass          http://localhost:8080/realms/master/;
            proxy_set_header    Host               $host;
            proxy_set_header    X-Real-IP          $remote_addr;
            proxy_set_header    X-Forwarded-For    $proxy_add_x_forwarded_for;
            proxy_set_header    X-Forwarded-Host   $host;
            proxy_set_header    X-Forwarded-Server $host;
            proxy_set_header    X-Forwarded-Port   $server_port;
            proxy_set_header    X-Forwarded-Proto  $scheme;
            allow 10.xxx.xxx.xxx;
            allow 10.xxx.xxx.xxx/24;
            deny all;
        }

        location / {
            proxy_pass          http://localhost:8080/;
            proxy_set_header    Host               $host;
            proxy_set_header    X-Real-IP          $remote_addr;
            proxy_set_header    X-Forwarded-For    $proxy_add_x_forwarded_for;
            proxy_set_header    X-Forwarded-Host   $host;
            proxy_set_header    X-Forwarded-Server $host;
            proxy_set_header    X-Forwarded-Port   $server_port;
            proxy_set_header    X-Forwarded-Proto  $scheme;
        }
```

/etc/nginx/ssl_security.conf
```
ssl_protocols TLSv1.3 TLSv1.2;
ssl_session_cache shared:SSL:10m;
ssl_ciphers HIGH:!aNULL:!MD5:!3DES:!CAMELLIA:!AES128:!AES256-SHA:!AES256-SHA256:!AES256-GCM-SHA384:!AES256-CCM8:!AES256-CCM;
ssl_prefer_server_ciphers on;
add_header Strict-Transport-Security "max-age=31536000; includeSubDomains; preload" always;
add_header Content-Security-Policy "upgrade-insecure-requests" always;
ssl_certificate /etc/ssl/domain.com/domain.com.crt;
ssl_certificate_key /etc/ssl/domain.com/domain.com.key;
ssl_stapling on;
```

**Reply by soundsessential** [reply]
Ah gotcha, that's great info too. Thanks.

**Reply by psytester** [reply]
Possible additional extension:
Because the console output is spamming (depending on log level) the /var/log/messages and also journalctl -xe, I redirect stdout & stderr to a file:
```ExecStart=/opt/keycloak/bin/kc.sh start --auto-build 2>&1 >> /opt/logfiles/console.log```

**Reply by SpiderD555** [reply]
> Are you using SSL certs and if so, where do you save these to? I seem to have had trouble with having Lets Encrypt certs save to their default location and then have a non root user get access to them.
> 
Thanks for diving into this topic, it seems [I have encountered the same problem as you](https://github.com/keycloak/keycloak/issues/15534).
I have also created my own [workaround](https://github.com/keycloak/keycloak/issues/15534#issuecomment-1327421141) (which is basically setting correct file permissions, so keycloak can read certificate and key files).
I hope it will help you (or someone else reading this topic).
